NSspi/Contexts/ImpersonationHandle.cs

```diff
@@ -29,6 +31,14 @@ internal ImpersonationHandle( ServerContext server )
             this.disposed = false;
         }
 
+        /// <summary>
+        /// Set the current thread security context to the impersonated identity
+        /// </summary>
+        public void SetThreadIdentity()
+        {
+            Thread.CurrentPrincipal = new WindowsPrincipal(WindowsIdentity.GetCurrent(TokenAccessLevels.AllAccess));

```

Comment:
The call to RevertImpersonate will unwind this because the underlying token will be closed and .NET will just revert on it's own.